src/test/test.cpp

```diff
@@ -425,4 +521,166 @@ for_all_versions_except(std::vector<uint32> const& versions, Application& app,
+{
+    LOG_ERROR(DEFAULT_LOG, "{}", msg);
+    throw std::runtime_error(msg);
+}
+
+static std::string
+getCurrentTestContext()
+{
+    assertThreadIsMain();
+    std::string ctxStr;
+    bool first = true;
+    for (auto const& tc : TestContextListener::sTestCtx)
+    {
+        if (!first)
+        {
+            ctxStr += ", ";
+        }
+        first = false;
+        ctxStr +=
+            fmt::format("File '{}', Test '{}'", tc.lineInfo.file, tc.name);
+    }
+    for (auto const& sc : TestContextListener::sSectCtx)
+    {
+        ctxStr += fmt::format(", Section '{}'", sc.name);
+    }
+    return ctxStr;
+}
+
+void
+recordOrCheckGlobalTestTxMetadata(TransactionMeta const& txmeta)
+{
+    if (gTestTxMetaMode == TestTxMetaMode::IGNORE)
+    {
+        return;
+    }
+    std::string ctx = getCurrentTestContext();
+    Hash h = xdrSha256(txmeta);
+    if (gTestTxMetaMode == TestTxMetaMode::RECORD)
+    {
+        gTestTxMetadata[ctx].first.emplace_back(h);
+    }
+    else
+    {
+        releaseAssert(gTestTxMetaMode == TestTxMetaMode::CHECK);
+        auto i = gTestTxMetadata.find(ctx);
+        if (i == gTestTxMetadata.end())
+        {
+            LOG_WARNING(DEFAULT_LOG, "Checking TxMeta for unknown context: {}",
+                        ctx);
+            LOG_WARNING(DEFAULT_LOG,
+                        "Consider re-recording expected TxMeta hashes");
+            return;
+        }
+        std::vector<Hash>& vec = i->second.first;
+        size_t& n = i->second.second;
+        if (n >= vec.size())
+        {
+            LOG_WARNING(DEFAULT_LOG,
+                        "Unexpected tx #{} while checking TxMeta hashes in {}",
+                        n, ctx);
+            LOG_WARNING(DEFAULT_LOG,
+                        "Consider re-recording expected TxMeta hashes");
+            return;
+        }
+        Hash& expected = vec.at(n);
+        if (expected == h)
+        {
+            LOG_DEBUG(DEFAULT_LOG, "Found expected TxMeta: {}",
+                      binToHex(expected));
+            ++n;
+        }
+        else
+        {
+            logErrAndThrow(fmt::format("Mismatched TxMeta hash: expected {}, "
+                                       "got {}, at tx #{}, in {}",
+                                       binToHex(expected), binToHex(h), n,
+                                       ctx));
+        }
+    }
+}
+
+static void
+loadTestTxMeta(std::string const& path)
+{
+    std::ifstream in(path);
+    if (!in)
+    {
+        logErrAndThrow(fmt::format("Failed to open {}", path));
+    }
+    size_t n = 0;
+    in.exceptions(std::ios::failbit | std::ios::badbit);
+    while (in.good() && in.peek() == '#')
+    {
+        char c;
+        in.get(c); // consume '#'
+        in.get(c); // consume ' '
+        if (c != ' ')
+        {
+            break;
+        }
+        std::string name;
+        std::getline(in, name);
+        std::vector<Hash> hashes;
+        while (in.good() && in.peek() == '\t')
```

Comment:
can we maybe use the `Json` class that we have and avoid rolling our own parser? That would limit the chance that we have a mismatch between the load and save code (note: I am not talking about cereal, which could also work, just plain `Json`)

Reply:
Again I was aiming for legibility (especially in diffs) but JSON is plausible enough, will change.